corehq/apps/registration/views.py

```diff
@@ -534,11 +534,17 @@ def confirm_domain(request, guid=''):
 def eula_agreement(request):
     if request.method == 'POST':
         current_user = CouchUser.from_django_user(request.user)
-        new_agreement = LicenseAgreement(type="End User License Agreement", version=str(EulaMixin.CURRENT_VERSION))
-        new_agreement.signed = True
-        new_agreement.date = datetime.utcnow()
-        new_agreement.user_ip = get_ip(request)
-        current_user.eulas.append(new_agreement)
+        if current_user.eula and not current_user.eula.signed:
```

Comment:
nit: my read of `EulaMixin.eula` is that it will always return a eula object, creating and appending one if necessary.  I think that'd make this conditional behave weirdly, as if there isn't a current eula, it'd create one (unsigned), add it to the list, then the second half of the conditional would return `False`, so this'd go to the `else` statement, which creates and adds another new eula.

As an aside, that method (that you didn't add) is a great example of the importance of [Command Query Separation](https://en.wikipedia.org/wiki/Command–query_separation), especially in a `@property` where it's not clear from accessing it that it's performing a query, let alone a modification.

> “Functions should either do something or answer something, but not both.”

Reply:
That might mean you can get away without the `if/else` stuff at all, depending on how we want to handle each of the cases where:

* A eula exists and is signed
* A eula exists but is unsigned
* A eula does not exist

Reply:
>as if there isn't a current eula, it'd create one (unsigned), add it to the list, then the second half of the conditional would return False

In testing the second half (`not current_user.eula.signed`) returns `True` and the just created eula is updated and saved. That leaves just one `eula` in `eulas`.
As you say, there are several scenarios to think about. I tested for these:
- user.eulas == [] 
- user.eula.date is signed, but older than a year
- user.eula is not signed
- user.eula version is old

It's a bit of an edge case, but if there is a pre-existing eula this check seems to handle it. It also happens to deal with no eula existing,  which is kind of a byproduct of not having Command Query Separation in this property.

Should reworking the `eula` `@property` should be a followup PR?

Reply:
So is this correct then?

* There is no existing eula of the current version - create and sign one
* There is an existing eula, but it's unsigned - sign it
* There is an existing, signed eula - create and sign another one

That all sounds reasonable to me

Reply:
yep, that's correct.